core/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/ShardsLimitAllocationDecider.java

```diff
@@ -145,20 +116,20 @@ public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAl
                 indexShardCount++;
             }
         }
-        // Subtle difference between the `canAllocate` and `canRemain` is that
-        // this checks > while canAllocate checks >=
-        if (clusterShardLimit > 0 && nodeShardCount > clusterShardLimit) {
+        // when it is allocate, checks >=
+        // when it is remain, checks >
+        if (clusterShardLimit > 0 && ((nodeShardCount > clusterShardLimit) || (isAllocate && nodeShardCount == clusterShardLimit))) {
```

Comment:
This is okay, but not what I had in mind. My thought was `canAllocate` and `canRemain` could pass two-parameter predicates to a common method `doDecide` as you've named it) and use those predicates for the decision. I think it will be clearer.